### ISSUE TITLE
`IndexError` choosing receptor reference atoms coming from `ligand_2_ref_idxs`

### ISSUE THREAD

**Issue by slochower** [Psivant/femto#15]
Hi, thanks again for putting this together. I'm taking a look at some internal test cases and we're seeing an issue where we get an `IndexError` when choosing receptor reference atoms. This is coming from the backbone/ligand distance matrix calculation.

Tracing this backwards, inside of `septop._setup._setup_system`, I see that `ligand_2_ref_idxs` are offset by the number of atoms in `ligand_1`:

```python
        ligand_2_ref_idxs = tuple(i + len(ligand_1.atoms) for i in ligand_2_ref_idxs)
```

At this point, `ligand_2_ref_idxs` can contain indices that are larger than the number of atoms in ligand 2. That is, if ligand 1 has 10 atoms and ligand 2 has 10 atoms and the `ligand_2_ref_idxs` starts at `[5, 6, 7]` (from `femto.fe.reference.select_ligand_idxs`), it will end up as `[15, 16, 17]`.

Then, when `select_receptor_idxs(...)` is called for the ligand 2, and then `_filter_receptor_atoms(...)` is called, the third argument is `ligand_ref_idxs[0]`. In the hypothetical example, this will be called with `15` for ligand 2. When the distances are calculated:

```python
    distances = scipy.spatial.distance.cdist(
        backbone.xyz[0, rigid_backbone_idxs, :], ligand.xyz[0, [ligand_ref_idx], :]
    )
```

`ligand` (which is ligand 2 in our example) is indexed directly by `ligand_ref_idx` -- which is 15 in our example -- even though the total length of `ligand.xyz` would be 10. This leads to an `IndexError`.

It seems like the intended way to index using an already-offset `ligand_ref_idx` would be in the combined ligand topology. But, oddly, we have run other systems without running into this issue, so I'm questioning whether my understanding of the code is correct. An obvious resolution would be to un-offset the index for ligand 2 at this point, but I'm not certain I have a full grasp of what's going on. Can you comment @SimonBoothroyd on if you've seen this before and if your indexing with the offset index is intentional?

**Comment by slochower** [comment]
Looking a little further, I think this bug only occurs if you go down the "alternate receptor reference atoms for ligand 2" path and if you avoid that, then you only index into ligand 1. It's this second function call that causes the problem.


```python
        receptor_ref_idxs_1 = femto.fe.reference.select_receptor_idxs(
            receptor, ligand_1, ligand_1_ref_idxs
        )
        receptor_ref_idxs_2 = receptor_ref_idxs_1

        if ligand_2 is not None and not femto.fe.reference.check_receptor_idxs(
            receptor, receptor_ref_idxs_1, ligand_2, ligand_1_ref_idxs
        ):
            _LOGGER.info("selecting alternate receptor reference atoms for ligand 2")

            # Calling `selector_receptor_idxs` with ligand 2 is the issue.
            receptor_ref_idxs_2 = femto.fe.reference.select_receptor_idxs(
                receptor, ligand_2, ligand_2_ref_idxs
            )
```

**Comment by slochower** [comment]
It seems like there could be two issues. I think this

```python
femto.fe.reference.check_receptor_idxs(receptor, receptor_ref_idxs_1, ligand_2, ligand_1_ref_idxs)
```

Should actually be
```python
# Reverse the offset
_ligand_2_ref_idxs = tuple(i - len(ligand_1.atoms) for i in ligand_2_ref_idxs)
femto.fe.reference.check_receptor_idxs(receptor, receptor_ref_idxs_1, ligand_2, _ligand_2_ref_idxs)
```

And then this
```python
receptor_ref_idxs_2 = femto.fe.reference.select_receptor_idxs(receptor, ligand_2, ligand_2_ref_idxs)
```

Should actually be something like
```python
# Reverse the offset
_ligand_2_ref_idxs = tuple(i - len(ligand_1.atoms) for i in ligand_2_ref_idxs)
receptor_ref_idxs_2 = femto.fe.reference.select_receptor_idxs(receptor, ligand_2, _ligand_2_ref_idxs)
```
I will test.

**Comment by SimonBoothroyd** [comment]
thanks so much for looking into this @slochower! I think @ajsilveira is planning to take a look at this and your fix!

**Comment by SimonBoothroyd** [comment]
fixed by #16
